### PR TITLE
Mxnet parser for Qnn dialect

### PR DESCRIPTION
**Pull request by shoubhik**
In this parser, I am including the changes needed for QNN dialect for Mxnet. It has been tested with most of the open-source Mxnet models.

### CODE REVIEW

**Comment by shoubhik** [comment]
@anijain2305 , @jackwish can you take a look at this.

File: python/tvm/relay/frontend/mxnet.py [hunk 1]
```diff
@@ -1075,6 +1098,377 @@ def _mx_cond(inputs, attrs, subgraphs):
     return ret
 
 
+def _qnn_mx_contrib_quantize(inputs, attrs):
```
Comment (liangfu) [inline python/tvm/relay/frontend/mxnet.py]: Can we put functions with `_qnn_mx` prefix into mxnet_qnn_op_utils.py file as well?

File: python/tvm/relay/frontend/mxnet.py [hunk 2]
```diff
@@ -14,12 +14,14 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
-# pylint: disable=invalid-name, import-self, len-as-condition, no-else-return
+# pylint: disable=invalid-name, import-self, len-as-condition, no-else-return, too-many-lines
```
[the remark this answers is not in the document]
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: This particular one is about the file, where pylint complains about too many lines in the file.

File: python/tvm/relay/frontend/mxnet_qnn_op_utils.py [hunk 3]
```diff
@@ -21,52 +21,95 @@
 
 import numpy as np
 from tvm import relay
-from tvm.relay.qnn.op.qnn import dequantize
+from tvm.relay.qnn.op.qnn import quantize, dequantize
 
-zero_centered_uint8_quantized_range = np.float32(255)
-zero_centered_int8_quantized_range = np.float32(127)
+zero_centered_uint8_quantized_range = np.float32(255.5)
```
[the remark this answers is not in the document]
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet_qnn_op_utils.py]: Added reference.

File: python/tvm/relay/frontend/mxnet.py [hunk 4]
```diff
@@ -180,6 +190,17 @@ def _mx_conv2d(inputs, attrs):
...
     new_attrs["data_layout"] = data_layout
     new_attrs["kernel_layout"] = kernel_layout
+    return new_attrs
+
+def _mx_conv2d(inputs, attrs):
+    kernel_size = attrs.get_int_tuple("kernel")
+    data_layout = attrs.get_str("layout", "NCHW")
+    if len(kernel_size) != 2:
+        raise tvm.error.OpAttributeInvalid(
+            'Non 1D or 2D kernels are not supported for operator Convolution')
```
[the remark this answers is not in the document]
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: Fixed the message

File: python/tvm/relay/frontend/mxnet.py (hunk 1 above)
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: Renamed the ops so that they no longer have _mx_ prefix in them.

File: python/tvm/relay/frontend/mxnet.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (masahi) [reply python/tvm/relay/frontend/mxnet.py]: actually we do have support for 1d and 3d conv in topi  now.

[the remark this answers is not in the document]
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: This condition seems to be from an older version of mxnet. It makes more sense to put this change in a separate PR after more extensive testing.

**Comment by masahi** [comment]
it would be great if you could an test case on end to end network translation like in our tflite test case.

File: src/relay/qnn/op/convolution.cc [hunk 5]
```diff
@@ -57,7 +57,7 @@ bool QnnConv2DRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
   CHECK(IsScalarType(types[2], DataType::Int(32)));    // input_zero_point
   CHECK(IsScalarType(types[3], DataType::Int(32)));    // kernel_zero_point
   CHECK(IsScalarType(types[4], DataType::Float(32)));  // input_scale
-  CHECK(IsScalarType(types[5], DataType::Float(32)));  // kernel_scale
+  // AssignType(types[1], DataType::Float(32), data->shape[axis], reporter);  // kernel_scale
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply src/relay/qnn/op/convolution.cc]: Handled in a separate PR - https://github.com/apache/incubator-tvm/pull/4732

File: python/tvm/relay/frontend/mxnet.py [hunk 6]
```diff
@@ -1033,6 +1054,7 @@ def _mx_contrib_fifo_buffer(inputs, attrs):
     new_attrs['axis'] = attrs.get_int('axis')
     return _op.nn.fifo_buffer(*inputs, **new_attrs)
 
+
```
Comment (jackwish) [inline python/tvm/relay/frontend/mxnet.py]: not needed?

File: python/tvm/relay/frontend/mxnet.py [hunk 7]
```diff
@@ -1075,6 +1097,378 @@ def _mx_cond(inputs, attrs, subgraphs):
...
+    out_type = attrs.get_str('out_type')
+    if out_type == 'auto':
+        if attrs.has_attr('min_calib_range') and attrs.has_attr('max_calib_range'):
+            if attrs.get_float('min_calib_range') >= 0:
+                out_dtype = 'uint8'
+            else:
+                out_dtype = 'int8'
+    else:
+        out_dtype = out_type
+    if out_dtype not in {'int8', 'uint8'}:
```
Comment (jackwish) [inline python/tvm/relay/frontend/mxnet.py]: It seems that this assert checks `out_type` actually? If that is the case, can we move it to where we obtained the `out_type` value?

File: python/tvm/relay/frontend/mxnet.py [hunk 8]
```diff
@@ -1075,6 +1097,378 @@ def _mx_cond(inputs, attrs, subgraphs):
...
+        raise ValueError('Unsupported out_dtype: %s' % out_dtype)
+    min_calib_range = attrs.get_float('min_calib_range', 0.0)
+    max_calib_range = attrs.get_float('max_calib_range', 0.0)
```
Comment (jackwish) [inline python/tvm/relay/frontend/mxnet.py]: That's interesting, if there is not `min/max_calib_range` provided, we are expecting the range to be `[0.0, 0.0)`? I am not very familiar with MxNet calibration method, but it appears to me that the *detault* min/max can derive from `out_dtype`?

File: python/tvm/relay/frontend/mxnet_qnn_op_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jackwish) [reply python/tvm/relay/frontend/mxnet_qnn_op_utils.py]: > Can you explain why are we adding an extra 0.5 here?

I guess this is related to the rounding policy and the quantization arithmetic?

File: python/tvm/relay/frontend/mxnet.py [hunk 9]
```diff
@@ -44,8 +58,9 @@
     "relu"   : _op.nn.relu
 }
 
+
 def _mx_fully_connected(inputs, attrs):
-    import mxnet as mx
+    import mxnet as mx #pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: running the pylint locally i was getting this warning.

File: python/tvm/relay/frontend/mxnet.py (hunk 8 above)
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: You are correct. The zeros should not be there. In fact there min and max operators. i'll fix this.

File: python/tvm/relay/frontend/mxnet.py (hunk 7 above)
Comment (shoubhik) [reply python/tvm/relay/frontend/mxnet.py]: out_dtype  can be auto too. in which case we infer it based on the min and max values.

**Comment by shoubhik** [comment]
> it would be great if you could add an test case on end to end network translation like in our tflite test case.

I have created a post on discuss on how we can do this - https://discuss.tvm.ai/t/use-mxnet-mkldnn-distribution-in-ci-instead-of-stock-mxnet-distribution/5474

**Comment by anijain2305** [comment]
With the latest commit, we have good accuracy for all MxNet-MKLDNN quantized models

@tmoreau89 @yzhliu @tqchen @jackwish @FrozenGene @liangfu 

  | MxNet | TVM | Degradation
-- | -- | -- | --
  | Top1 | Top5 | Top1 | Top5 | Top1 | Top5
Resnet18_v1 | 69.76 | 89.02 | 69.85 | 89.09 | -0.09 | -0.07
Resnet50_v1 | 76.13 | 92.6 | 75.9 | 92.66 | 0.23 | -0.06
Resnet50_v1b | 76.66 | 92.6 | 76.45 | 92.57 | 0.21 | 0.03
Resnet101_v1 | 77.13 | 93.06 | 77 | 93.06 | 0.13 | 0
Resnet152_v2 | 75.99 | 92.52 | 75.32 | 92.26 | 0.67 | 0.26
Inception-V3 | 77.84 | 93.52 | 77.28 | 93.32 | 0.56 | 0.2
Inception-BN | 71.96 | 90.38 | 71.79 | 90.25 | 0.17 | 0.13
MobileNetV1 | 71.27 | 90.09 | 71.13 | 90.16 | 0.14 | -0.07
MobileNetV2 | 70.35 | 89.45 | 70.19 | 89.5 | 0.16 | -0.05

**Comment by shoubhik** [comment]
@tmoreau89 @yzhliu @tqchen @jackwish @FrozenGene @liangfu @vinx13 @anijain2305  I want to discuss this PR from testability POV. Mxnet suggests to use MKLDNN as backend for quantization. The operators for MKLDNN quantization in Mxnet is different from stock Mxnet quantization operators. Also, for some ops the implementation of these ops are also different.
Due to this PR #4753 and #4764, it is not possible to merge the Mxnet-MKLDNN installation in TVM docker. One potential solution is to test stock quantization operators but that would be extra work that may not be too useful in future. All the testing we have done and the edge cases we have covered are for the MKLDNN operators.
My suggestion is at this point we check in the parser code along with the the test scripts we have for benchmarking the QNN networks. When MKLDNN issues are fixed and we can safely upgrade we can add test cases at that point.

**Comment by anijain2305** [comment]
Given the constraints, I am ok with MxNet-parser code w/o any MxNet-MKLDNN CI testing, as long as we have a test script linked in the PR, that somebody can locally use to test the changes.

Once Mxnet-MKLDNN CI/accuracy issues are resolved, we can re-write that script as a test. 

Is that ok - @tqchen @icemelon9

**Comment by FrozenGene** [comment]
> @tmoreau89 @yzhliu @tqchen @jackwish @FrozenGene @liangfu @vinx13 @anijain2305 I want to discuss this PR from testability POV. Mxnet suggests to use MKLDNN as backend for quantization. The operators for MKLDNN quantization in Mxnet is different from stock Mxnet quantization operators. Also, for some ops the implementation of these ops are also different.
> Due to this PR #4753 and #4764, it is not possible to merge the Mxnet-MKLDNN installation in TVM docker. One potential solution is to test stock quantization operators but that would be extra work that may not be too useful in future. All the testing we have done and the edge cases we have covered are for the MKLDNN operators.
> My suggestion is at this point we check in the parser code along with the the test scripts we have for benchmarking the QNN networks. When MKLDNN issues are fixed and we can safely upgrade we can add test cases at that point.

I am OK with this.

**Comment by shoubhik** [comment]
@tqchen @yzhliu  I think the code can be merged now.

File: tests/python/frontend/mxnet/test_forward.py [hunk 10]
```diff
@@ -28,68 +28,173 @@
...
+                                                       quantized_dtype=quantized_dtype,
+                                                       logger=logger)
+        return cqsym, cqarg_params, _aux_params
+
+    ctx = mx.cpu(0)
+    x = mx.random.uniform(-1.0, 1.0, shape=data_shape, ctx=ctx, dtype='float32')
+    _, arg_params, aux_params = get_mxnet_output(mx_symbol, x)
+    quantize_api = quantize_with_old_api if use_old_mxnet_quantization_api() else quantized_with_new_api
+
+
```
Comment (masahi) [inline tests/python/frontend/mxnet/test_forward.py]: remove this white spaces

**Comment by masahi** [comment]
@shoubhik I'll have a look and merge. I'm interested in this PR.

**Comment by masahi** [comment]
thanks @shoubhik @anijain2305 @jackwish @FrozenGene @liangfu this is merged.

**Comment by masahi** [comment]
> My suggestion is at this point we check in the parser code along with the the test scripts we have for benchmarking the QNN networks

Where are that test scripts? I want to run them @shoubhik @anijain2305
